README.md
# aidbox-monitoring

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2016 FIXME

Distributed under the Eclipse Public License either version 1.0 or (at
your option) any later version.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-03-25
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-03-25
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/aidbox-monitoring/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/aidbox-monitoring/compare/0.1.0...0.1.1


test/aidbox_monitoring/core_test.clj
(ns aidbox-monitoring.core-test
  (:require [clojure.test :refer :all]
            [aidbox-monitoring.parser :refer :all]
            [aidbox-monitoring.core :refer :all]))

(def nginx-access
  [["87.249.45.131 - - [29/Mar/2016:12:33:08 +0000] \"GET /boxes HTTP/1.1\" 200 2325 \"https://aidbox.io/ui\" \"Mozilla/5.0  (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/49.0.2623.87 Chrome/49.0.2623.87 Safari/537.36\""
    {:ip "87.249.45.131" :date "29/Mar/2016:12:33:08 +0000" :method "GET" :url "/boxes" :statuscode "200" :bytessent "2325" :refferer "https://aidbox.io/ui" :useragent "Mozilla/5.0  (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/49.0.2623.87 Chrome/49.0.2623.87 Safari/537.36"} ]

   ["87.249.45.131 - - [29/Mar/2016:12:33:08 +0000] \"POST /boxes/12/ui HTTP/1.1\" 400 2325 \"https://aidbox.io/ui/users\" \"Mozilla/5.0  (X11; Linux x86_64) AppleWebKit/537.36\""
    {:ip "87.249.45.131" :date "29/Mar/2016:12:33:08 +0000" :method "POST" :url "/boxes/12/ui" :statuscode "400" :bytessent "2325" :refferer "https://aidbox.io/ui/users" :useragent "Mozilla/5.0  (X11; Linux x86_64) AppleWebKit/537.36"} ] ])

(def nginx-error
  [["2016/03/23 16:11:57 [error] 30770#0: *3 connect() failed (111: Connection refused) while connecting to upstream, client: 87.249.45.131, server: ~^(?<sub>.+)\\.dev.aidbox.io$, request: \"GET /favicon.ico HTTP/1.1\", upstream: \"http://127.0.0.1:8080/favicon.ico\", host: \"dev.aidbox.io\", referrer: \"http://dev.aidbox.io/ui\"" 
  {:date "2016/03/23 16:11:57" :status "error" :errorcode "30770#0" :errmessage "*3 connect() failed (111: Connection refused) while connecting to upstream, client: 87.249.45.131, server: ~^(?<sub>.+)\\.dev.aidbox.io$, request: \"GET /favicon.ico HTTP/1.1\", upstream: \"http://127.0.0.1:8080/favicon.ico\", host: \"dev.aidbox.io\", referrer: \"http://dev.aidbox.io/ui\"" }]

  ["2016/03/23 16:25:49 [notice] 1447#0: signal process started"
   {:date "2016/03/23 16:25:49" :status "notice" :errorcode "1447#0" :errmessage "signal process started"}]])

(def app-log
  [["16-03-30 13:20:49.051 INFO [worker-1] default simply.core || :get   /test-exception"
    {:date "16-03-30 13:20:49.051" :status "INFO" :worker "worker-1" :context "default" :logger "simply.core" :message ":get   /test-exception"  } ]
   ["16-03-30 13:20:49.068 WARN [worker-1] default simply.db || DEPRECATED db/q-one*" 
    {:date "16-03-30 13:20:49.068" :status "WARN" :worker "worker-1" :context "default" :logger "simply.db" :message "DEPRECATED db/q-one*"  }]
   ["16-03-30 14:26:26.448 ERROR [worker-1] default simply.core || java.lang.Exception: Test exception
     at dashboard.core$$test_exception.invoke (core.clj:58)
        clojure.lang.Var.invoke (Var.java:379)
        simply.core$dispatch$fn__23818.invoke (core.clj:103)"
    {:date "16-03-30 14:26:26.448" :status "ERROR" :worker "worker-1" :context "default" :logger "simply.core" :message "java.lang.Exception: Test exception
     at dashboard.core$$test_exception.invoke (core.clj:58)
        clojure.lang.Var.invoke (Var.java:379)
        simply.core$dispatch$fn__23818.invoke (core.clj:103)"  } ] ])

(deftest nginx-access-log
  (testing "Nginx access log"
    (doseq [[req res] nginx-access]
      (is (= (parse-nginx-access req)  res))))
  (testing "Nginx error log"
    (doseq [[req res] nginx-error]
      (is (= (parse-nginx-error req)  res))))
  (testing "App log"
    (doseq [[req res] app-log]
      (is (= (parse-app req)  res)))))

(run-tests)


project.clj
(defproject aidbox-monitoring "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :pedantic? true
  :warn true
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :main aidbox-monitoring.core
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [clj-http "2.1.0"]
                 [org.clojure/tools.cli "0.3.3"]
                 [clojurewerkz/elastisch "2.2.1"]
                 [ch.qos.logback/logback-classic "1.1.3"]
                 [com.niquola/riemann "0.2.11-SNAPSHOT" :exclusions [org.slf4j/slf4j-api  org.jsoup/jsoup clj-http]]
                 [clj-time "0.11.0"]
                 [cheshire "5.5.0"]
                 [org.elasticsearch/elasticsearch "1.7.4"]
                 [org.apache.lucene/lucene-expressions "4.10.4"]])


src/aidbox_monitoring/dev.clj
(ns aidbox-monitoring.dev
  (:require
   [aidbox-monitoring.elastic :as es]
   [riemann.config :as cfg]
   [riemann.client :as cli]
   [riemann.streams :as s]
   [riemann.transport.sse :as sse]
   [riemann.logging :as logging]
   [riemann.time :as time]
   [riemann.core :as c]))

(defonce server (atom nil))
(defonce tcp (riemann.transport.tcp/tcp-server {:port 5555}))

(defn stop []
  (when-let [srv @server]
    (c/stop! srv)
    (reset! server nil)))


(defn reload [cfg]
  (reset! server (c/transition! (c/core) (assoc cfg :services [tcp])))
  "reloaded")

(defn start [cfg]
  (time/start!)
  (es/es-connect)
  (stop)
  (reload cfg)
  "started")

(def elastic
  (s/rollup 1 3 (es/es-index "kiries" :timestamping :day :index "logstash")))

(defn process-stream [e]
  #_(doseq [[k v] e] (println k ": " (pr-str  v)))
  (cond
    (= (:service e) "nginx-access-log") e
    :else e))

(def stream (s/smap #'process-stream elastic))

(comment
  (if @server
    (reload {:streams  [stream]})
    (start {:streams  [stream]}))
  (stop))


src/aidbox_monitoring/core.clj
(ns aidbox-monitoring.core
  (:require
   [aidbox-monitoring.elastic :as es]
   [aidbox-monitoring.parser :refer :all]
   [riemann.config :as cfg]
   [riemann.client :as cli]
   [riemann.streams :as s]
   [riemann.slack :refer [slack]]
   [riemann.transport.sse :as sse]
   [riemann.logging :as logging]
   [riemann.time :as time]
   [riemann.core :as c]))

(defonce server (atom nil))
(defonce tcp (riemann.transport.tcp/tcp-server {:port 5555 :host "0.0.0.0"}))

(def slack-credentials
  {:account "???"
   :token "???" })

(def slacker
  (slack slack-credentials  {:username "incoming-webhook" :channel "#aidbox"
                             :text "ERROR" }))

(defn stop []
  (when-let [srv @server]
    (c/stop! srv)
    (reset! server nil)))


(defn reload [cfg]
  (reset! server (c/transition! (c/core) (assoc cfg :services [tcp])))
  "reloaded")

(defn start [cfg]
  (time/start!)
  (es/es-connect)
  (stop)
  (reload cfg)
  "started")

(def elastic
  (s/rollup 1 3 (es/es-index "kiries" :timestamping :day :index "logstash")))

(defn process-stream [e]
  #_(doseq [[k v] e] (println k ": " (pr-str  v)))
  (cond
    (= (:service e) "nginx_access")
      (merge e (parse-nginx-access (:description e)) )
    (= (:service e) "nginx_error")
      (merge e (parse-nginx-error (:description e)) )
    (= (:service e) "aidbox")
      (let [e (merge e (parse-app (:description e)))]
        (when (= (:status e) "ERROR")
          (slacker e))
        e)


    :else e))

(def stream (s/smap #'process-stream elastic))

(defn -main [& args]
  (start {:streams  [stream]}))

(comment
  (if @server
    (reload {:streams  [stream]})
    (start {:streams  [stream]}))
  (stop))


src/aidbox_monitoring/elastic.clj
(ns aidbox-monitoring.elastic
  (:require [cheshire.core :as json]
            [clj-time.format]
            [clj-time.core]
            [clj-time.coerce]
            [clojure.edn :as edn]
            [clojure.java.io :as io]
            [clojure.tools.logging :as log]
            [clojurewerkz.elastisch.rest.bulk :as eb]
            [clojurewerkz.elastisch.rest :as esr]
            [riemann.streams :as streams]))

;; See http://www.elastichq.org/

(defn make-index-timestamper [index period]
  (let [formatter (clj-time.format/formatter (str "'" index "'"
                                  (cond
                                   (= period :day)
                                   "-YYYY.MM.dd"
                                   (= period :hour)
                                   "-YYYY.MM.dd.HH"
                                   (= period :week)
                                   "-YYYY.MM.dd.ww"
                                   (= period :month)
                                   "-YYYY.MM"
                                   (= period :year)
                                   "-YYYY")))]
    (fn [date]
      (clj-time.format/unparse formatter date))))

(def ^{:private true} format-iso8601
  (clj-time.format/with-zone (clj-time.format/formatters :date-time)
    clj-time.core/utc))

(defn ^{:private true} iso8601 [event-s]
  (clj-time.format/unparse format-iso8601
                           (clj-time.coerce/from-long (* 1000 event-s))))

(defn ^{:private true} safe-iso8601 [event-s]
  (try (iso8601 event-s)
    (catch Exception e
      (log/warn "Unable to parse iso8601 input: " event-s)
      (clj-time.format/unparse format-iso8601 (clj-time.core/now)))))

(defn ^{:private true} stashify-timestamp [event]
  (-> (if-not (get event "@timestamp")
        (assoc event "@timestamp"
               (if-let [isotime (:isotime event)]
                 isotime
                 (if-let [time (:time event)]
                   (safe-iso8601 (long time))
                   (clj-time.format/unparse format-iso8601 (clj-time.core/now)))))
         event)
      (dissoc :isotime :time :ttl)))

(defn ^{:private true} edn-safe-read [v]
  (try
    (edn/read-string v)
    (catch Exception e
      (log/warn "Unable to read supposed EDN form with value: " v)
      v)))

(defn ^{:private true} massage-event [event]
  (into {}
        (for [[k v] event
              :when v]
          (cond
           (= (name k) "_id") [k v]
           (.startsWith (name k) "_")
           [(.substring (name k) 1) (edn-safe-read v)]
           :else
           [k v]))))

(defn ^{:private true} elastic-event [event massage]
  (let [e (-> event
              stashify-timestamp)]
    (if massage
      (massage-event e)
      e)))

(defn  riemann-to-elasticsearch [events massage]
  (->> [events]
       flatten
       (remove streams/expired?)
       (map #(elastic-event % massage))))

(defonce es-connection (atom nil))

(defn es-connect
  "Connects to the ElasticSearch node.  The optional argument is a url
  for the node, which defaults to `http://localhost:9200`.  This must
  be called before any es-* functions can be used."
  [& argv]
  (reset! es-connection
          (esr/connect (or (first argv) "http://localhost:9200"))))


(defn es-index
  "A function which takes a sequence of events, and indexes them in
  ElasticSearch.  It will set the `_type` field of the event to the
  value of `doc-type`, which tells ES which mapping to use for the
  event.

  The :index argument defaults to \"logstash\" for Kibana
  compatability.  It's the root ES index name that this event will be
  indexed within.

  The :timestamping argument, default to :day and it controls the time
  range component of the index name.  Acceptable values
  are :hour, :day, :week, :month and :year.

  Events will be massages to conform to Kibana expections.  This means
  that the `@timestamp` field will be set if not found, based on the
  `time` field of the event.  The `ttl` field will be removed, as it's
  internal to Riemann.  Lastly, any fields starting with an `_` will
  have their value parsed as EDN.
"
  [doc-type & {:keys [index timestamping massage]
               :or {index "logstash"
                    massage true
                    timestamping :day}}]
  (let [index-namer (make-index-timestamper index timestamping)]
    (fn [events]
      (let [esets (group-by (fn [e]
                              (index-namer
                               (clj-time.format/parse format-iso8601
                                                      (get e "@timestamp"))))
                            (riemann-to-elasticsearch events massage))]
        (doseq [index (keys esets)]
          (let [raw (get esets index)
                bulk-create-items
                (interleave (map #(if-let [id (get % "_id")]
                                    {:create {:_type doc-type :_id id}}
                                    {:create {:_type doc-type}})
                                 raw)
                            raw)]
            (when (seq bulk-create-items)
              (try
                (let [res (eb/bulk-with-index @es-connection index bulk-create-items)
                      total (count (:items res))
                      succ (filter #(= 201 (get-in % [:create :status])) (:items res))
                      failed (filter :error (:items res))]
                  (when-not (empty? failed)
                    (log/error "Failed: " failed)))
                (catch Exception e
                  (log/error "Unable to bulk index:" e))))))))))

(defn ^{:private true} resource-as-json [resource-name]
  (json/parse-string (slurp (io/resource resource-name))))


(defn ^{:private true} file-as-json [file-name]
  (try
    (json/parse-string (slurp file-name))
    (catch Exception e
      (log/error "Exception while reading JSON file: " file-name)
      (throw e))))


(defn load-index-template
  "Loads the file into ElasticSearch as an index template."
  [template-name mapping-file]
  (esr/put @es-connection
           (esr/index-template-url @es-connection template-name)
           {:body (file-as-json mapping-file)}))

(comment
  (es-connect)
  @es-connection
  )


src/aidbox_monitoring/parser.clj
(ns aidbox-monitoring.parser
  (:require [clojure.string :as str]))

(def nginx-access-keys
  [:ip :date :method :url :statuscode :bytessent :refferer :useragent ])
(def nginx-access-regex 
  #"^(\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}) - - \[(.+?)\] \"(GET|POST|PUT|DELETE) (.+) HTTP/1\.1\" (\d{1,3}) (\d+) \"(.+)\" \"(.+)\"")

(def nginx-error-keys
  [:date :status :errorcode :errmessage])
(def nginx-error-regex 
  #"^(.+? .+?) \[(.+?)\] (.+?): (.+)$")

(def app-keys
  [:date :status :worker :context :logger :message]  )
(def app-regex 
  #"^(.+? .+?) (.+?) \[(.+?)\] (.+?) (.+?)$")


(defn mk-parser [keys regex]
  (fn [msg] 
    (zipmap keys (rest (re-find regex msg)))))

(def parse-nginx-access
  (mk-parser nginx-access-keys nginx-access-regex))

(def parse-nginx-error 
  (mk-parser nginx-error-keys nginx-error-regex))

(def app-log-parser
  (mk-parser app-keys app-regex))

(defn parse-app [message]
  (let [[status msg] (str/split message #"\|\|")]
    (assoc (app-log-parser (str/trim status))
           :message (str/trim msg))) )
  
(def na-log
    "87.249.45.131 - - [29/Mar/2016:12:33:08 +0000] \"GET /boxes HTTP/1.1\" 200 2325 \"https://aidbox.io/ui\" \"Mozilla/5.0  (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/49.0.2623.87 Chrome/49.0.2623.87 Safari/537.36\"")

(def ne-log 
  "2016/03/23 16:11:57 [error] 30770#0: *3 connect() failed (111: Connection refused) while connecting to upstream, client: 87.249.45.131, server: ~^(?<sub>.+)\\.dev.aidbox.io$, request: \"GET /favicon.ico HTTP/1.1\", upstream: \"http://127.0.0.1:8080/favicon.ico\", host: \"dev.aidbox.io\", referrer: \"http://dev.aidbox.io/ui\"")

(def a-log
  "16-03-30 14:26:26.448 ERROR [worker-1] default simply.core || java.lang.Exception: Test exception
     at dashboard.core$$test_exception.invoke (core.clj:58)
        clojure.lang.Var.invoke (Var.java:379)
        simply.core$dispatch$fn__23818.invoke (core.clj:103)
  ")

(comment 
  (parse-nginx-access na-log)
  (parse-nginx-error ne-log)
  (parse-app a-log))
